src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp

```diff
@@ -1,85 +1,179 @@
+        ai_config_.shoot_or_pass_play_config().short_pass_threshold();
+
+    const Polygon pass_area_polygon =
+        Polygon::fromSegment(Segment(passer_point, receiver_point), 0.5);
+
+    // calculate a polygon that contains the receiver and passer point, and checks if the
+    // ball is inside it. if the ball isn't being passed to the receiver then we should
+    // abort
+    if ((receiver_point - passer_point).length() >= short_pass_threshold)
     {
-        // Always reduce number of defenders if we have more bots on the field than
-        // the enemy team, so that we take advantage of the scoring opportunity
-        // while the enemy team is down a robot
-        num_defenders = 1;
+        if (!contains(pass_area_polygon, ball_position))
+        {
+            return true;
+        }
     }
-    else if (event.common.num_tactics <= 2)
+
+    // distance between robot and ball is too far, and it's not in flight,
+    // i.e. team might still have possession, but kicker/passer doesn't have control over
+    // ball
+    const double ball_velocity = event.common.world_ptr->ball().velocity().length();
+    const double ball_shot_threshold =
+        ai_config_.shoot_or_pass_play_config().ball_shot_threshold();
+    const double min_distance_to_pass =
+        ai_config_.shoot_or_pass_play_config().min_distance_to_pass();
+
+    if ((ball_velocity < ball_shot_threshold) &&
+        ((ball_position - passer_point).length() > min_distance_to_pass))
     {
-        // Set number of defenders to 0 so we don't get stuck defending and never
-        // attempt to score
-        num_defenders = 0;
+        return true;
     }
-    else
+
+    return receiver_tactic_->done();
+}
+
+void OffensePlayFSM::attack(const Update& event)
+{
+    PriorityTacticVector tactics = {{attacker_tactic_, receiver_tactic_}, {}};
+
+    if (event.common.num_tactics > 2)
     {
-        num_defenders = 2;
-    }
+        std::vector<Point> existing_receiver_positions = {
+            best_pass_with_rating_.pass.receiverPoint()};
 
-    num_shoot_or_pass = event.common.num_tactics - num_defenders;
+        updateOffensivePositioningTactics(event.common.world_ptr,
+                                          event.common.num_tactics - 2,
+                                          existing_receiver_positions);
+
+        tactics[1].insert(tactics[1].end(), offensive_positioning_tactics_.begin(),
+                          offensive_positioning_tactics_.end());
+    }
 
-    setTactics(event, num_shoot_or_pass, num_defenders);
+    event.common.set_tactics(tactics);
 }
 
-void OffensePlayFSM::setupDefensiveStrategy(const Update& event)
+void OffensePlayFSM::receive(const Update& event)
 {
-    setTactics(event, 0, event.common.num_tactics);
+    PriorityTacticVector tactics = {{receiver_tactic_}, {}};
+
+    std::vector<Point> existing_receiver_positions = {
+        best_pass_with_rating_.pass.receiverPoint()};
+
+    updateOffensivePositioningTactics(
```

Comment:
you should check if `event.common.num_tactics` is > 0 before this operation (unsigned ints will underflow)